### PR TITLE
find-domains query perf fixes

### PR DESCRIPTION
**Pull request by shrugs**
## Also Included: Bridged-Resolver Canonicality Fix

scope creep against the original perf focus, but found while inspecting the perf-test index: only 4.9M of 13.5M Domains were canonical — all mainnet, with the entire ~8.6M Basenames + Lineanames subtree non-canonical. root cause: `apps/ensindexer/src/plugins/unigraph/handlers/ensv1/ENSv1Registry.ts` unconditionally called `handleSubregistryUpdated` and `handleRegistryCanonicalDomainUpdated` on every NewOwner, clobbering the two pointers that `handleBridgedResolverChange` owns (Domain.subregistryId on the L1 origin, Registry.canonicalDomainId on the L2 target). either side's first chain-local subname event broke the bidirectional agreement.

fix: guard both calls with two new SDK predicates `isBridgeOriginDomain` / `isBridgedTargetRegistry` (added to `packages/ensnode-sdk/src/shared/protocol-acceleration/is-bridged-resolver.ts`, reusing the cached config list). on bridge endpoints we leave the pointer alone; non-bridged subnames behave identically to before.

intended outcome (post-reindex): mainnet `linea.eth`/`base.eth`.subregistryId stays on the L2 bridged Registry, the L2 Registry's canonicalDomainId stays on the mainnet origin Domain, agreement holds, cascade canonicalizes all ~8.6M Lineanames + Basenames. mainnet-registered subnames directly under `linea.eth`/`base.eth` (`bridge.linea.eth`, `devconnect.linea.eth`, the handful that exist) drop out of canonicality — consistent with bridged-resolver semantics where resolution goes through L2.

---

## Reviewer Focus (Read This First)

two things worth scrutiny:

1. **cursor format change.** NAME-ordered cursors now store the 256-char prefix of `canonical_name` instead of the full name (see `truncateNameForCursor`). cursors minted by the prior deploy will not decode-compare correctly under the new code — clients with in-flight pagination state will need to restart from page 1.
2. **canonical filter semantics.** `DomainsWhere.canonical` is now `typeof === "boolean"`-gated: both `true` and `false` produce an `eq(canonical, ...)` filter; only `null`/`undefined` is "no filter". `AccountDomainsWhereInput.canonical` lost its `defaultValue: false` — clients omitting that arg will now receive all domains regardless of canonicality (was: effectively no-op-then-everything, since `false` used to be a no-op). `Query.domains` no longer hardcodes `canonical: true` — the required `name` filter already implies it via `canonical_name IS NOT NULL ⟺ canonical = true`.

## Problem & Motivation

three slow paths against the mainnet perf-test index (13.5M domains, 4.9M canonical):

- `get-domain-by-interpreted-name("eth")`: 2.69s. namegraph walk's recursive CTE left-joined `domain_resolver_relations` on `domain_id` alone, but the PK leads with `(chain_id, address)` so the join had no usable index.
- `Domain.subdomains(.eth)` (3.5M children): 15.26s+. registry-scoped browse had no composite index for `WHERE registry_id = X ORDER BY canonical_name LIMIT N`; planner fell back to filter + sort.
- broad-prefix `Query.domains` typeahead (e.g. `'a%'` ordered by NAME): planner was thrown by the hardcoded `canonical = true` predicate. removing it lets the planner use the GIN trigram path cleanly.

separately, `find-domains` had a layered "filter-by-X" abstraction (`base-domain-set → filter-by-name → filter-by-canonical → ...`) that materialized intermediate CTEs and made the SELECT shape opaque to the planner.

## What Changed (Concrete)

1. **new indexes** in `packages/ensdb-sdk/src/ensindexer-abstract/`:
   - `domain_resolver_relations(domain_id)` (secondary lookup off the PK)
   - `domains(registry_id, left(canonical_name, 256), id)` (composite for registry-scoped NAME browse)
2. **find-domains resolver rewrite** (`find-domains-resolver.ts` and helpers): layered filter-by-X pattern collapsed into a single flat `resolveFindDomains` with a compound `WHERE`. SELECT projects only `id` + a single conditional `registration_value` column. NAME/DEPTH order values are read back from the dataloader-hydrated `Domain` instead of carried in a separate map. layer files deleted.
3. **Domain.subdomains** now forward-walks via `parent.subregistryId` instead of reverse-walking via `registry.canonicalDomainId`. when `subregistryId` is null, short-circuits to a new `EMPTY_CONNECTION` helper (in `connection-helpers.ts`).
4. **cursor encoding**: NAME-ordered cursors are truncated to `CANONICAL_NAME_SORT_PREFIX` (256) chars at encode time via `truncateNameForCursor`. the filter-side `left(${cursor.value}::text, 256)` re-application is gone; cursor value is compared directly against the index expression.
5. **canonical filter semantics**: `DomainsWhere.canonical` is `typeof === "boolean"`-gated, so both `true` and `false` filter. `AccountDomainsWhereInput.canonical` drops `defaultValue: false`; description rewritten to clarify the three states. `Query.domains` drops the hardcoded `canonical: true` — the required name filter does the work.
6. **cursorFilter NULL branches** collapsed 4 → 2 by hoisting `op` and `idCmp`.
7. **VERSION_TO_DOMAIN_TYPE** value side typed against `(typeof ensIndexerSchema.domainType.enumValues)[number]` instead of a hand-rolled literal union.
8. **DomainsNameFilterValue** replaced with `typeof DomainsNameFilter.\$inferInput` inline at point-of-use.
9. **schema.graphql** regenerated to reflect `AccountDomainsWhereInput.canonical` no longer defaulting.

## Design & Planning

no design doc. all work was done incrementally against a live perf-test index (`ensindexer_perf_unigraph_pa`, `mainnet`, `unigraph + protocol-acceleration`) with EXPLAIN ANALYZE driving each index decision.

an early version of this PR added two partial indexes `WHERE canonical = true` to support `Query.domains`'s hardcoded `canonical = true` predicate. follow-up analysis showed that the NAME partial was *harmful* in the broad-prefix typeahead case (planner mis-selected it and scanned 4M rows; 11s actual). dropping the hardcoded predicate and the partials together is strictly better — the GIN trigram + existing hash indexes handle the remaining shapes, and the only regression (`'a%'` DEPTH-order: 5.5ms → 686ms) is a single-keystroke edge case.

alternative considered: enforce `InterpretedLabel` byte-length cap (255-byte DNS-encodable) and materialize a length-capped canonical name column, removing the need for the `LEFT(...)` expression index. documented in the `CANONICAL_NAME_SORT_PREFIX` JSDoc as the cleaner long-term shape; deferred because it requires changes to indexing-time semantics.

- planning artifacts: none
- reviewed / approved by: n/a (solo)

## Self-Review

ran `/simplify` (three review agents: reuse, quality, efficiency) and applied findings.

- bugs caught: SELECT projected `canonicalName` + `canonicalDepth` on every row regardless of orderBy; narrowed to id + conditional registration column. planner mis-selection on partial NAME index under broad-prefix ILIKE.
- logic simplified: `__orderValue` map was redundant for NAME/DEPTH — those values live on the loaded Domain. `getOrderValueFromResult` deleted.
- naming / terminology improved: `truncateNameForCursor`, `EMPTY_CONNECTION`, `getDefaultOrder`. `DOMAINS_DEFAULT_ORDER` constant extracted.
- dead or unnecessary code removed: layered filter-by-X files (`base-domain-set.ts`, `filter-by-*.ts`, `with-ordering-metadata.ts`, `layers/index.ts`) deleted; `getOrderValueFromResult` deleted; verbose `DomainsWhere` JSDoc trimmed; narrating comments ("Compound WHERE…", "build order clauses…") removed; the two partial `WHERE canonical = true` indexes removed.

## Cross-Codebase Alignment

- search terms used: `resolveFindDomains`, `DomainsWhereInput`, `canonical = true`, `setFilterCondition`, `paginateBy`, `byCanonicalDepth`
- reviewed but unchanged: `find-events-resolver.ts` (similar `\$dynamic()` + conditional join shape — left independent), `Registry.parents` (separate path), `Account.domains` callsite (now relies on the new semantics), `Query.domains` callsite (no longer needs the predicate)
- deferred alignment: lift `setFilterCondition` from `find-events-resolver.ts` into a shared module so `nameCondition` can reuse its `eq`/`in` halves. cross-file refactor; out of scope here.

## Downstream & Consumer Impact

- public APIs affected:
  - `AccountDomainsWhereInput.canonical` no longer defaults — described in §Reviewer Focus
  - cursor format change — described in §Reviewer Focus
- docs updated: `AccountDomainsWhereInput.canonical` GraphQL description rewritten; schema.graphql regenerated.
- naming worth calling out: `EMPTY_CONNECTION` is a module-level singleton, not a factory. pothos relay reads `edges`/`pageInfo` without mutating, so the shared instance is safe — but if any future plugin starts mutating connection returns, this would leak across requests.

## Testing Evidence

- 98 ensapi unit tests pass; monorepo typecheck + `pnpm lint:ci` clean.
- live EXPLAIN ANALYZE against perf-test mainnet (post-fix, cold/warm):
  - `get-domain-by-interpreted-name("eth")`: 2.69s → 0.836ms
  - `Domain.subdomains(.eth)` (3.5M children): 15.26s → 19-22ms warm / 140ms cold
  - `Query.domains` `name.eq` ordered NAME: 0.78ms (hash index)
  - `Query.domains` `name.starts_with: "vitalik"` ordered NAME: 16.6ms / DEPTH: 1.4ms (GIN trigram + bitmap)
  - `Query.domains` worst-case `name.starts_with: "a"` (815k matches) ordered NAME: 1.8s / DEPTH: 686ms — single-keystroke typeahead edge case, sub-second
- known gaps: integration tests (`pnpm test:integration:ci`) not run for the post-merge state. cursor-encoding change has no dedicated unit test (existing helpers tests cover only `isEffectiveDesc`).
- what reviewers reason about manually: that pothos relay's `pageInfo`/`edges` consumers don't mutate the returned object (relevant to `EMPTY_CONNECTION` singleton).

## Scope Reductions

- lift `setFilterCondition` into a shared helper for both `find-events` and `find-domains` — separate refactor PR
- enforce `InterpretedLabel` byte-length cap to avoid the `LEFT(...)` expression index entirely — needs indexing-time changes; documented as future work in `CANONICAL_NAME_SORT_PREFIX` JSDoc
- convert `EMPTY_CONNECTION` to a factory function — only matters if a future pothos plugin mutates connection results; trivial to switch later

## Risk Analysis

- risk: existing cursors won't pagination-compare correctly. mitigation: cursors are typically session/request-scoped; clients will restart pagination on the deploy boundary.
- risk: `Account.domains` default behavior changes — clients that relied on `canonical: false` as a no-op now get a filter applied. mitigation: schema description rewritten; `defaultValue` removed so omitting the arg gives the most permissive answer (all domains). a client explicitly passing `canonical: false` is the only one affected, and the new behavior matches the field's name.
- risk: single-keystroke typeahead (`name.starts_with: "a"`) DEPTH-order is now ~700ms (was ~5ms with the dropped partial). worth mentioning to UI clients that they should debounce or require ≥2 chars; all narrower prefixes are sub-20ms.
- named owner: @shrugs

## Pre-Review Checklist (Blocking)

- [x] I reviewed every line of this diff and understand it end-to-end
- [x] I'm prepared to defend this PR line-by-line in review
- [x] I'm comfortable being the on-call owner for this change
- [x] Relevant changesets are included (or explicitly not required)

### CODE REVIEW

File: packages/ensdb-sdk/src/ensindexer-abstract/ensv2.schema.ts [hunk 1]
```diff
@@ -338,10 +338,26 @@ export const domain = onchainTable(
   }),
   (t) => ({
     byType: index().on(t.type),
-    byRegistry: index().on(t.registryId),
     bySubregistry: index().on(t.subregistryId).where(sql`${t.subregistryId} IS NOT NULL`),
     byOwner: index().on(t.ownerId),
     byLabelHash: index().on(t.labelHash),
+    // Composite for `(registry_id, label_hash)` lookups (namegraph walk in
```
Comment (lightwalker-eth) [inline packages/ensdb-sdk/src/ensindexer-abstract/ensv2.schema.ts:344]: @shrugs Appreciate your help to keep ENSDb docs site content in sync 👍 Thanks

File: apps/ensapi/src/omnigraph-api/lib/find-domains/find-domains-resolver-helpers.ts [hunk 2]
```diff
@@ -1,23 +1,62 @@
...
+export const CANONICAL_NAME_SORT_PREFIX = 256;
+
+/**
+ * Truncate a `canonicalName` to the cursor / index prefix length. Used when writing the cursor
+ * value for NAME orderings — callers slice once at encode time so the encoded cursor stays small
+ * (long names can hit thousands of characters) and `cursorFilter` can compare directly against
+ * the index expression without re-applying `left(...)` per row.
+ */
+export function truncateNameForCursor(name: string | null): string | null {
+  return name === null ? null : name.slice(0, CANONICAL_NAME_SORT_PREFIX);
```
Comment (lightwalker-eth) [inline apps/ensapi/src/omnigraph-api/lib/find-domains/find-domains-resolver-helpers.ts]: @shrugs Hey there's an important Unicode issue to fix here. JavaScript's string functions don't do what is intuitive for them to do.

Please see here for related background info: https://github.com/namehash/ensnode/issues/1987

https://github.com/namehash/namekit/blob/main/packages/ens-utils/src/ensname.ts#L329-L365

File: apps/ensapi/src/omnigraph-api/schema/domain.ts [hunk 3]
```diff
@@ -230,11 +225,13 @@ DomainInterfaceRef.implement({
...
-        const domains = withOrderingMetadata(named);
+        if (!parent.subregistryId) return EMPTY_CONNECTION;
 
-        return resolveFindDomains(context, { domains, order, defaultOrder, ...connectionArgs });
+        return resolveFindDomains(context, {
+          where: { ...where, registryId: parent.subregistryId },
+          order,
+          ...connectionArgs,
+        });
       },
```
[the remark this answers is not in the document]
Comment (lightwalker-eth) [reply apps/ensapi/src/omnigraph-api/schema/domain.ts:235]: @shrugs Looks like a comment to investigate more deeply.

File: apps/ensapi/src/omnigraph-api/schema/domain.ts [hunk 4]
```diff
@@ -230,11 +225,13 @@ DomainInterfaceRef.implement({
...
-        const { named, defaultOrder } = filterByName(base, where?.name ?? null);
-        const domains = withOrderingMetadata(named);
+        if (!parent.subregistryId) return EMPTY_CONNECTION;
 
-        return resolveFindDomains(context, { domains, order, defaultOrder, ...connectionArgs });
+        return resolveFindDomains(context, {
+          where: { ...where, registryId: parent.subregistryId },
+          order,
+          ...connectionArgs,
+        });
```
[the remark this answers is not in the document]
Comment (lightwalker-eth) [reply apps/ensapi/src/omnigraph-api/schema/domain.ts:234]: @shrugs Looks like a comment to investigate more deeply.

File: apps/ensapi/src/omnigraph-api/lib/find-domains/find-domains-resolver-helpers.ts [hunk 5]
```diff
@@ -1,23 +1,54 @@
...
```
[the remark this answers is not in the document]
Comment (shrugs) [reply apps/ensapi/src/omnigraph-api/lib/find-domains/find-domains-resolver-helpers.ts]: Good catch — fixed. `truncateNameForCursor` now iterates code points via `[...name].slice(0, 256).join("")` to match Postgres `left(text, 256)` character semantics, so surrogate pairs at the boundary no longer split.

File: apps/ensapi/src/omnigraph-api/lib/find-domains/find-domains-resolver-helpers.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (shrugs) [reply apps/ensapi/src/omnigraph-api/lib/find-domains/find-domains-resolver-helpers.ts]: Fixed — `truncateNameForCursor` is now code-point-safe (`[...name].slice(0, 256).join("")`).

File: apps/ensapi/src/omnigraph-api/lib/find-domains/find-domains-resolver-helpers.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (shrugs) [reply apps/ensapi/src/omnigraph-api/lib/find-domains/find-domains-resolver-helpers.ts]: Intentional — documented on `CANONICAL_NAME_SORT_PREFIX` in `find-domains-resolver-helpers.ts`. Domains with names differing only past character 256 tie-break by `id` rather than by the full `canonical_name`. The tradeoff exists because the index expression `left(canonical_name, 256)` is what makes registry-scoped NAME-ordered browse possible at all (raw `canonical_name` ordering falls back to a full scan + sort against a 22GB table). The cleaner long-term shape — enforce an `InterpretedLabel` byte-length cap upstream so the index can use the full materialized column — is captured as deferred work in the same JSDoc.

File: apps/ensapi/src/omnigraph-api/schema/domain.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (shrugs) [reply apps/ensapi/src/omnigraph-api/schema/domain.ts:234]: Intentional forward-walk: `Domain.subdomains` returns every Domain whose `registry_id` is this Domain's `subregistryId`. The previous reverse-walk variant (filter `registry.canonicalDomainId = parent.id` then join) ran in 15s+ on `.eth` (3.5M children) — the planner had no usable stats for the JOIN-derived predicate. The forward-walk is 5ms cold via the new `(registry_id, left(canonical_name, 256), id)` composite. The semantic difference (which @greptile-apps also flagged as a P2) only matters in the narrow window between a `subregistry_id` update on the parent and the corresponding `canonical_domain_id` flip on the registry. In a steady-state canonical tree the two definitions match. `Domain.parent` continues to use the reverse-walk via `registry.canonical_domain_id`.

File: apps/ensapi/src/omnigraph-api/schema/domain.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (shrugs) [reply apps/ensapi/src/omnigraph-api/schema/domain.ts:235]: Confirmed — intentional perf tradeoff. The forward-walk via `subregistryId` is the only shape that keeps `Domain.subdomains(.eth)` sub-100ms; the bidirectional reverse-walk ran in 15s+ because the planner couldn't estimate selectivity for the `registry.canonical_domain_id = parent.id` JOIN-derived predicate. The transient-window divergence only matters between a parent's `subregistryId` update and the registry's reciprocal `canonicalDomainId` flip, which is bounded by a single handler transaction in practice.

**Review by lightwalker-eth — APPROVED** [review]
@shrugs Great to see these enhancements 👍 Reviewed and shared feedback. Please feel welcome to merge when ready!

File: apps/ensapi/src/omnigraph-api/lib/find-domains/find-domains-resolver-helpers.ts [hunk 6]
```diff
@@ -1,23 +1,62 @@
...
+ *
+ * The btree per-tuple max is ~2712 bytes; with `registry_id` and `id` consuming ~240 bytes of
+ * that, ~2400 bytes remain for the prefix expression. 256 chars × max 4-byte UTF-8 codepoint =
+ * 1024 bytes, well under the limit and within the realm of reasonable name lengths (mainnet avg
+ * is ~126). Queries MUST sort by this same expression for the planner to use the index for
+ * ordered scan; raw `canonical_name` ORDER BY falls back to a full scan + sort.
+ *
+ * An alternative solution is to redefine InterpretedLabel to enforce a maximum byte length of 255 before
+ * being truncated into an Encoded LabelHash — this mirros a name's resolvability (must be dns-encodable)
+ * and allows us to avoid storing spam names. Then we'd also have to produce an b-tree indexed
```
[the remark this answers is not in the document]
Comment (shrugs) [reply apps/ensapi/src/omnigraph-api/lib/find-domains/find-domains-resolver-helpers.ts]: Fixed — "mirros" -> "mirrors", "an b-tree" -> "a b-tree-indexed".

[the remark this answers is not in the document]
Comment (shrugs) [reply apps/ensapi/src/omnigraph-api/lib/find-domains/find-domains-resolver-helpers.ts]: Fixed.

File: packages/ensdb-sdk/src/ensindexer-abstract/ensv2.schema.ts [hunk 7]
```diff
@@ -338,10 +338,26 @@ export const domain = onchainTable(
...
+    // to bound the index tuple under btree's per-tuple max (~2712 bytes): 256 chars × max 4-byte
+    // UTF-8 = 1024 bytes, leaving ample room for the registry_id and id columns. Names beyond
+    // 256 chars (currently <0.0001% of mainnet) collide on the truncated prefix and tie-break by
+    // id; this is acceptable since such names are invariably spam. Callers MUST sort by the same
+    // expression for the planner to use this index for ordered scan.
+    byRegistryAndCanonicalNameLeft: index().on(
+      t.registryId,
+      sql`left(${t.canonicalName}, 256)`,
+      t.id,
+    ),
```
[the remark this answers is not in the document]
Comment (shrugs) [reply packages/ensdb-sdk/src/ensindexer-abstract/ensv2.schema.ts:360]: Added `.changeset/find-domains-perf-indexes.md` (`@ensnode/ensdb-sdk` patch) describing the three new indexes.

File: packages/ensdb-sdk/src/ensindexer-abstract/ensv2.schema.ts [hunk 8]
```diff
@@ -338,10 +338,26 @@ export const domain = onchainTable(
...
     byType: index().on(t.type),
-    byRegistry: index().on(t.registryId),
     bySubregistry: index().on(t.subregistryId).where(sql`${t.subregistryId} IS NOT NULL`),
     byOwner: index().on(t.ownerId),
     byLabelHash: index().on(t.labelHash),
+    // Composite for `(registry_id, label_hash)` lookups (namegraph walk in
+    // get-domain-by-interpreted-name.ts). The leading `registry_id` column also serves
+    // `WHERE registry_id = X` lookups via prefix scan.
+    byRegistryAndLabelHash: index().on(t.registryId, t.labelHash),
+
```
[the remark this answers is not in the document]
Comment (shrugs) [reply packages/ensdb-sdk/src/ensindexer-abstract/ensv2.schema.ts:348]: Added `.changeset/find-domains-perf-indexes.md` (`@ensnode/ensdb-sdk` patch).

File: apps/ensapi/src/omnigraph-api/lib/connection-helpers.ts [hunk 9]
```diff
@@ -40,3 +41,19 @@ export const paginateByInt = (
...
+  connection: async () => ({
+    edges: [],
+    pageInfo: {
+      hasNextPage: false,
+      hasPreviousPage: false,
+      startCursor: null,
+      endCursor: null,
+    },
+  }),
+});
```
[the remark this answers is not in the document]
Comment (shrugs) [reply apps/ensapi/src/omnigraph-api/lib/connection-helpers.ts:59]: Acknowledged and deferred — captured under §Scope Reductions in the PR description as "convert EMPTY_CONNECTION to a factory function — only matters if a future pothos plugin mutates connection results; trivial to switch later." Current pothos relay reads edges/pageInfo without mutating, so the singleton is safe today.

File: apps/ensapi/src/omnigraph-api/lib/find-domains/find-domains-resolver.ts [hunk 10]
```diff
@@ -172,18 +242,22 @@ export function resolveFindDomains(
...
-
+            const __orderValue: DomainOrderValue = (() => {
+              switch (orderBy) {
+                case "NAME":
+                  return truncateNameForCursor(domain.canonicalName);
+                case "DEPTH":
+                  return domain.canonicalDepth;
+                case "REGISTRATION_TIMESTAMP":
+                case "REGISTRATION_EXPIRY":
+                  return registrationValueById?.get(domain.id) ?? null;
```
[the remark this answers is not in the document]
Comment (shrugs) [reply apps/ensapi/src/omnigraph-api/lib/find-domains/find-domains-resolver.ts]: Good catch — replaced the `?.` with an explicit invariant guard. When `orderBy` is `REGISTRATION_*`, `registrationValueById` is guaranteed populated; the new code throws if that ever is not the case rather than silently returning `null` for a missing id.

**Comment by shrugs** [comment]
@greptile review

**Comment by shrugs** [comment]
@greptile review
